File: lgpio/python/lcd_1inch44.py
# Seengreat 1.44 Inch LCD Display demo
# Author(s):Andy Li from Seengreat

import os
import sys
import spidev
import time
from gpiozero import *
import numpy as np

# define the key button pin in BCM number
RST_PIN  = 27
DC_PIN   = 25
BL_PIN   = 24
class LCD_1Inch44():
    def __init__(self):
        # gpio init
#         wpi.wiringPiSetup()
#         wpi.pinMode(DC_PIN, wpi.OUTPUT)  # D/C pin
        self.dc = DigitalOutputDevice( DC_PIN,active_high = True,initial_value =False)       
#         wpi.pinMode(RST_PIN, wpi.OUTPUT)  # reset pin
        self.rst = DigitalOutputDevice( RST_PIN,active_high = True,initial_value =False)
#         wpi.pinMode(BL_PIN, wpi.OUTPUT)  # back light control pin
        self.bl = PWMOutputDevice(BL_PIN,frequency = 1000)
        self.bl.value = 0.9
#         wpi.softPwmCreate(BL_PIN, 0, 100)
#         wpi.softPwmWrite(BL_PIN, 90)
        # spi init
        self.bus = 0
        self.dev = 0
        self.spi_speed = 8000000
        self.spi = spidev.SpiDev()
        self.spi.open(self.bus, self.dev)
        self.spi.max_speed_hz = self.spi_speed
        self.spi.mode = 0b00
        # self.spi.xfer([4000000, 10, 8])
        # define width and height of the 2 inch lcd
        self.w = 128
        self.h = 128 
        
    def write_cmd(self, cmd):
        """write command"""
#         wpi.digitalWrite(DC_PIN,wpi.LOW)
        self.dc.off()
        self.spi.writebytes([cmd])
        
    def write_data(self, value):
        """write data"""
#         wpi.digitalWrite(DC_PIN,wpi.HIGH)
        self.dc.on()
        self.spi.writebytes([value])
        
    def reset(self):
        """reset the lcd"""
#         wpi.digitalWrite(RST_PIN,wpi.HIGH)
        self.rst.on()
        time.sleep(0.01)
#         wpi.digitalWrite(RST_PIN,wpi.LOW)
        self.rst.off()
        time.sleep(0.01)
#         wpi.digitalWrite(RST_PIN,wpi.HIGH)
        self.rst.on()
        time.sleep(0.01)
    def lcd_init(self):
        """lcd init..."""
        self.reset()
        
        self.write_cmd(0x01) #software reset
        time.sleep(0.02)
        
        self.write_cmd(0x11) #sleep out
        time.sleep(0.05)       
        # FRMCTR1:Frame Rate Control in normal mode
        self.write_cmd(0xB1)
        self.write_data(0x02)
        self.write_data(0x35)
        self.write_data(0x36)
        
        # FRMCTR2:Frame Rate Control in idle mode
        self.write_cmd(0xB2)
        self.write_data(0x02)
        self.write_data(0x35)
        self.write_data(0x36)
        
        # FRMCTR3:Frame Rate Control in Partial mode/full colors
        self.write_cmd(0xB3)
        self.write_data(0x02)
        self.write_data(0x35)
        self.write_data(0x36)
        self.write_data(0x02)
        self.write_data(0x35)
        self.write_data(0x36)
        
        # INVCTR:Displa INversion Control
        self.write_cmd(0xB4) 
        self.write_data(0x03)
        
        # PWCTR1:Power Control 1
        self.write_cmd(0xC0) 
        self.write_data(0xA2)
        self.write_data(0x02)
        self.write_data(0x84)
        
        self.write_cmd(0xC1) 
        self.write_data(0xC5)

        self.write_cmd(0xC2) 
        self.write_data(0x0D)
        self.write_data(0x00)

        self.write_cmd(0xC3) 
        self.write_data(0x8D)
        self.write_data(0x2A)
        
        self.write_cmd(0xC4) 
        self.write_data(0x8D)
        self.write_data(0xEE)

        self.write_cmd(0xC5) # VCOM
        self.write_data(0x0A)

        self.write_cmd(0x36) # MX,MY,RGB MODE
        self.write_data(0x68) #0xEC
        
        self.write_cmd(0xE0) 
        self.write_data(0x12)
        self.write_data(0x1C)
        self.write_data(0x10)
        self.write_data(0x18)
        self.write_data(0x33)
        self.write_data(0x2C)
        self.write_data(0x25)
        self.write_data(0x28)
        self.write_data(0x28)
        self.write_data(0x27)
        self.write_data(0x2F)
        self.write_data(0x3C)
        self.write_data(0x00)
        self.write_data(0x03)
        self.write_data(0x03)
        self.write_data(0x10)
        
        self.write_cmd(0xE1) 
        self.write_data(0x12)
        self.write_data(0x1C)
        self.write_data(0x10)
        self.write_data(0x18)
        self.write_data(0x2D)
        self.write_data(0x28)
        self.write_data(0x23)
        self.write_data(0x28)
        self.write_data(0x28)
        self.write_data(0x26)
        self.write_data(0x2F)
        self.write_data(0x3B)
        self.write_data(0x00)
        self.write_data(0x03)
        self.write_data(0x03)
        self.write_data(0x10)
        
        self.write_cmd(0xF0) 
        self.write_data(0x01)
        
        self.write_cmd(0xF6) 
        self.write_data(0x00)

        self.write_cmd(0x3A) 
        self.write_data(0x05)
        
        #self.write_cmd(0x11)
        #time.sleep(0.12)
        self.write_cmd(0x29)
        time.sleep(0.05)
        
    def set_windows(self, start_x, start_y, end_x, end_y):
        """display the windows of start point and end point"""
        # because register 36h = 0x68
        start_x += 1
        end_x += 1
        start_y += 2
        end_y += 2
        # set the start point and end point of x
        self.write_cmd(0x2A) 
        self.write_data(start_x >> 8)  # x address start
        self.write_data(start_x & 0xff)
        self.write_data(end_x >> 8)  # x address end
        self.write_data(end_x & 0xff)
        # set the start point and end point of y
        self.write_cmd(0x2B) 
        self.write_data(start_y >> 8)  # y address start
        self.write_data(start_y & 0xff)
        self.write_data(end_y >> 8)  # y address end
        self.write_data(end_y & 0xff)
        
        self.write_cmd(0x2C)
        
    def img_show(self, img):
        img_w, img_h = img.size
        
        image = np.asarray(img)
        pixel = np.zeros((self.w, self.h, 2), dtype = np.uint8)
        pixel[..., [0]] = np.add(np.bitwise_and(image[..., [0]], 0xf8), np.right_shift(image[..., [1]], 5))
        
        pixel[..., [1]] = np.add(np.bitwise_and(np.left_shift(image[..., [1]], 3), 0xe0), np.right_shift(image[..., [2]], 3))       
        
        pixel = pixel.flatten().tolist()
        
        self.set_windows(0, 0, 127, 127) # 36H=0XEC:2, 3, 129, 130; 36H=0X68:1, 2, 128, 129 
#         wpi.digitalWrite(DC_PIN,wpi.HIGH)
        self.dc.on()
        for i in range(0, len(pixel), 4096):
            self.spi.writebytes(pixel[i:i+4096])
                
    def clear(self):
        self.set_windows(0, 0, 127, 127) # 36H=0XEC:2, 3, 129, 130; 36H=0X68:1, 2, 128, 129 
#         wpi.digitalWrite(DC_PIN,wpi.HIGH)
        self.dc.on()
        buf = [0xFF]*(self.w*self.h*2)
        for i in range(0, len(buf), 4096):
            self.spi.writebytes(buf[i:i+4096])


    

File: lgpio/c/lcd_1inch44.c
/********************************
 *  Seengreat 1.44 Inch LCD Display demo
 *  Author(s):Andy Li from Seengreat
 * ********************/
 
#include <string.h>
#include <stdlib.h>		
#include "sg_lgpio.h"

#include "lcd_1inch44.h"

void Lcd_Gpio_Init(void)
{
	printf("Init gpio for lgpio\r\n");
    Lgpio_set_pinmode(PIN_RST,LGPIO_OUTPUT);
    Lgpio_set_pinmode(PIN_DC,LGPIO_OUTPUT);
    LG_PWM(PIN_BL,1000,90);//set screen backlight
    LG_SPI_Init(0, 0, 8000000);//spi0.0
}

/*******************************************************************************
function:
	Hardware reset
*******************************************************************************/
static void Lcd_Reset(void)
{
	//digitalWrite(PIN_CS, 1);
	delay(100);
    Lgpio_write_pin(PIN_RST, LGPIO_LOW);
	delay(100);
    Lgpio_write_pin(PIN_RST, LGPIO_HIGH);
	delay(100);
}

/*******************************************************************************
function:
		Write data and commands
*******************************************************************************/
static void Spi_write_cmd(u_char data)	 
{	
	//digitalWrite(PIN_CS, 0);
	Lgpio_write_pin(PIN_DC, LGPIO_LOW);
	//wiringPiSPIDataRW(0,&data,1);
	LG_SPI_writebyte(data);
}

static void Spi_write_data(u_char data) 
{	
	//digitalWrite(PIN_CS, 0);
	Lgpio_write_pin(PIN_DC, LGPIO_HIGH);
	LG_SPI_writebyte(data);
	//digitalWrite(PIN_CS,1);
}  

void Spi_write_word(u_int16_t data)
{
	u_char temp = 0;
	//digitalWrite(PIN_CS, 0);
	Lgpio_write_pin(PIN_DC, LGPIO_HIGH);
	temp = (data>>8) & 0xff;
	LG_SPI_writebyte(temp);
	temp = data & 0xff;
	LG_SPI_writebyte(temp);
	//digitalWrite(PIN_CS, 1);
}	  


/******************************************************************************
function:	
		Common register initialization
******************************************************************************/
void Lcd_Init(void)
{
	Lcd_Gpio_Init();
	Lcd_Reset();

	Spi_write_cmd(0x01);
	delay(20); 

	Spi_write_cmd(0x11); 
	delay(50); 

	Spi_write_cmd(0xB1);
	Spi_write_data(0x02);
	Spi_write_data(0x35);
	Spi_write_data(0x36);

	Spi_write_cmd(0xB2);
	Spi_write_data(0x02);
	Spi_write_data(0x35);
	Spi_write_data(0x36);

	Spi_write_cmd(0xB3);
	Spi_write_data(0x02);
	Spi_write_data(0x35);
	Spi_write_data(0x36);
	Spi_write_data(0x02);
	Spi_write_data(0x35);
	Spi_write_data(0x36);

	Spi_write_cmd(0xB4);
	Spi_write_data(0x03); 

	Spi_write_cmd(0xBB);
	Spi_write_data(0x1F);

	Spi_write_cmd(0xC0);
	Spi_write_data(0xA2);
	Spi_write_data(0x02);
	Spi_write_data(0x84);

	Spi_write_cmd(0xC1);
	Spi_write_data(0xC5);

	Spi_write_cmd(0xC2);
	Spi_write_data(0x0D);
	Spi_write_data(0x00);
	
	Spi_write_cmd(0xC3);
	Spi_write_data(0x8D);   
	Spi_write_data(0x2A);   

	Spi_write_cmd(0xC4);
	Spi_write_data(0x8D);
	Spi_write_data(0xEE);

	Spi_write_cmd(0xC5);
	Spi_write_data(0x0A); 

	Spi_write_cmd(0x36);
	Spi_write_data(0x68); 

	Spi_write_cmd(0xE0);
	Spi_write_data(0x12);
	Spi_write_data(0x1C);
	Spi_write_data(0x10);
	Spi_write_data(0x18);
	Spi_write_data(0x33);
	Spi_write_data(0x2C);
	Spi_write_data(0x25);
	Spi_write_data(0x28);
	Spi_write_data(0x28);
	Spi_write_data(0x27);
	Spi_write_data(0x2F);
	Spi_write_data(0x3C);
	Spi_write_data(0x00);
	Spi_write_data(0x03);
	Spi_write_data(0x03);
	Spi_write_data(0x10);

	Spi_write_cmd(0xE1);
	Spi_write_data(0x12);
	Spi_write_data(0x1C);
	Spi_write_data(0x10);
	Spi_write_data(0x18);
	Spi_write_data(0x2D);
	Spi_write_data(0x28);
	Spi_write_data(0x23);
	Spi_write_data(0x28);
	Spi_write_data(0x28);
	Spi_write_data(0x26);
	Spi_write_data(0x2F);
	Spi_write_data(0x3B);
	Spi_write_data(0x00);
	Spi_write_data(0x03);
	Spi_write_data(0x03);
	Spi_write_data(0x10);
	
	Spi_write_cmd(0xF0);
	Spi_write_data(0x01);

	Spi_write_cmd(0xF6);
	Spi_write_data(0x00);

	Spi_write_cmd(0x3A);
	Spi_write_data(0x05);
	

	Spi_write_cmd(0x29);
	delay(50);
}


//select the srceen area(from point(start_x,start_y) to point(end_x,end_y))
void Set_Window(u_int16_t start_x, u_int16_t start_y, u_int16_t end_x, u_int16_t  end_y)
{ 
	start_x += 1;
	end_x += 1;
	start_y += 2;
	end_y += 2;
	Spi_write_cmd(0x2a);
	Spi_write_data(start_x >>8);
	Spi_write_data(start_x & 0xff);
	Spi_write_data(end_x >> 8);
	Spi_write_data(end_x & 0xff);

	Spi_write_cmd(0x2b);
	Spi_write_data(start_y >>8);
	Spi_write_data(start_y & 0xff);
	Spi_write_data(end_y >> 8);
	Spi_write_data(end_y & 0xff);

	Spi_write_cmd(0x2C);
}


//set cursor to the point(x, y)
void Set_Cursor(u_int16_t x, u_int16_t y)
{ 
	x += 1;
	y += 2;
	Spi_write_cmd(0x2a);
	Spi_write_data(x >> 8);
	Spi_write_data(x);
	Spi_write_data(x >> 8);
	Spi_write_data(x);

	Spi_write_cmd(0x2b);
	Spi_write_data(y >> 8);
	Spi_write_data(y);
	Spi_write_data(y >> 8);
	Spi_write_data(y);

	Spi_write_cmd(0x2C);
}

//refresh the entire screen to color
void Lcd_Clear(u_int16_t color)
{
	u_int16_t i;
	u_int16_t image[LCD_W];
	for(i=0;i<LCD_W;i++)
	{
		image[i] = color>>8 | (color&0xff)<<8;
	}
	u_char *p = (u_char *)(image);
	Set_Window(0, 0, LCD_W-1, LCD_H-1);
	Lgpio_write_pin(PIN_DC, LGPIO_HIGH);
	for(i = 0; i < LCD_H; i++)
	{
		//wiringPiSPIDataRW(0, (unsigned char *)p, LCD_W*2);
		LG_SPI_write_bytes((unsigned char *)p, LCD_W*2);
	}
}


//display the image to whole screen 
void Display(u_int16_t *image)
{
	u_int16_t i;
	u_int16_t p[LCD_SIZE];
    memcpy(p,image,LCD_SIZE);
	Set_Window(0, 0, LCD_W-1, LCD_H-1);
	Lgpio_write_pin(PIN_DC, LGPIO_HIGH);
	for(i = 0; i < LCD_H; i++)
	{
		//wiringPiSPIDataRW(0, (unsigned char *)p+LCD_W*2*i, LCD_W*2);
		LG_SPI_write_bytes((unsigned char *)p+LCD_W*2*i, LCD_W*2);
	}
}

//set the point(x,y) to color
void Draw_Point(u_int16_t x, u_int16_t y, u_int16_t Color)
{
	Set_Cursor(x, y);
	Spi_write_word(Color); 	    
}

